### ISSUE TITLE
freelist.py 

### ISSUE THREAD

**Issue by Mineru98** [hyunwoongko/pecab#56]
https://github.com/hyunwoongko/pecab/blob/37a75f679e472126f632702b0fce889cfc0994c2/mecab/utils/freelist.py#L33

위 코드는 일단 기존에 (코드)[https://github.com/jeongukjae/python-mecab/blob/master/include/mecab/utils/freelist.h#L27]의 목적대로 동작하고 있지 않습니다. 제가 C++ 코드를 제대로 이해하지 못한 상태로 수정이 작업을 한 것 같습니다.

그런데 이전 커밋에서 작성된 (freelist.py)[https://github.com/hyunwoongko/pecab/blob/05a8e80b28/mecab/utils/freelist.py#L31]에서는 `IndexError: list index out of range` 에러에 대응할 수 있는 아이디어는 나올 필요가 있어 보입니다.

```python
if len(self.free_list) == 0:
    obj = self.free_list
    return obj
else:
    obj = self.free_list[self.li_][self.pi_]
    self.pi_ += 1
    return obj
```

이렇게 한다면 일단 테스트 코드에서는 동작하는걸 확인했습니다.

**Comment by Mineru98** [comment]
제가 새롭게 올린 코드에서 `obj = self.free_list[self.li_] = self.free_list[self.li_] + self.pi_` 코드 자체가 일단 기존에 동작해야 하는 방식이 아니라 제가 잘못 해석한 방향으로 코드가 작성이 되었습니다.
제일 문제는 `list`에  `int`를 덧셈하는 연산을 해서 에러코드가 잘 기억이 안나는데요. `list`, `int` 두개의 타입 간에는 덧셈 연산이 없다는 식의 에러가 생겼습니다.

**Comment by hyunwoongko** [comment]
list에 int 더하는게 포인터 연산 아닌가용?
그 만큼 주소 이동하는거요.

**Comment by Mineru98** [comment]
cpp 코드는 포인터 연산이지만, 파이썬에선 그렇게 표현이 되지 않으니까요.

**Comment by hyunwoongko** [comment]
네 그래서 해당 리스트에서 pi++ 만큼 이동한 자리에 있는 요소를 반환해주도록 제가 작성했던 것 같은데요.
그 부분이 문제인건가용? (제가 잘못짰을수도...)

**Comment by hyunwoongko** [comment]
음.. `alloc` 부분에서 접근을 하고 그 부분이 문제인가요?

**Comment by Mineru98** [comment]
네, Allocator 클래스에서 FreeList 클래스로 path_freelist 멤버가 생성자에서 size가 0으로 초기화되기 때문에 FreeList의 alloc 메소드에서 len 함수를 통해 길이가 0인 경우에는 따로 로직 처리를 해줘야 한다는게 제 의견입니다.

그런데 이런 경우 변환되는 값이 길이가 0이 아닌 경우와 다른 반환 값이 될 것이다보니 나중에 문제가 될 수도 있을거라 생각이 듭니다.

**Comment by hyunwoongko** [comment]
요 부분 맡아서 처리해주실 수 있으실까요?

**Comment by Mineru98** [comment]
일단 Allocator 클래스에서 사용하는데에는 문제 없도록 수정해서 커밋을 해뒀습니다. 이후에 혹시 문제가 발생할것 같아서 이슈를 남겨두긴합니다. 이슈 제목을 적당한거로 바꾸면 좋을것 같긴네요.
